drivers: crypto: Add support for STM32L4 AES accelerator

lucasdietrich opened the pull request:
The STM32L4 series features an AES accelerator with differences in the HAL API compared to other STM32 devices. This patch adds support for the STM32L4 AES accelerator in the STM32 crypto driver.

Key changes:

- Introduce a new device type `st,stm32l4-aes` to represent the STM32L4 AES accelerator variant. Add the device tree binding file `st,stm32l4-aes.yaml` and update `stm32l462.dtsi` accordingly.
- Modify `crypto_stm32.c` to handle the differences in the HAL API:
  - Use specific HAL functions for encryption and decryption (e.g., `HAL_CRYP_AESECB_Encrypt`) when building for STM32L4 AES.
  - Adjust the initialization and configuration structures accordingly.
  - Replace `copy_reverse_words` with `copy_words_adjust_endianness` to handle endianness conversion only when necessary.

The driver has been successfully tested with the `crypto` sample on a custom board based on the STM32L462RE.

---

Related to issue #78482

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -159,24 +188,23 @@ static int crypto_stm32_ecb_decrypt(struct cipher_ctx *ctx,
 		return -EINVAL;
 	}
 
-	ret = do_decrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf);
+	ret = do_aes(ctx, hal_ecb_decrypt_op, pkt->in_buf, pkt->in_len, pkt->out_buf);
 	if (ret == 0) {
 		pkt->out_len = 16;
 	}
 
 	return ret;
 }
 
-static int crypto_stm32_cbc_encrypt(struct cipher_ctx *ctx,
-				    struct cipher_pkt *pkt, uint8_t *iv)
+static int crypto_stm32_cbc_encrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt, uint8_t *iv)
```
Avoid style changes if not required, it confuses review w/o additional review.
If really required, make it a dedicated commit.

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -75,17 +115,22 @@ static int do_encrypt(struct cipher_ctx *ctx, uint8_t *in_buf, int in_len,
 
 	k_sem_take(&data->device_sem, K_FOREVER);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
 	status = HAL_CRYP_SetConfig(&data->hcryp, &session->config);
 	if (status != HAL_OK) {
 		LOG_ERR("Configuration error");
 		k_sem_give(&data->device_sem);
 		return -EIO;
 	}
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
+	/* Device is initialized from the configuration in the encryption/decryption function
+	 * called bellow. */
+	memcpy(&data->hcryp.Init, &session->config, sizeof(session->config));
+#endif
 
-	status = HAL_CRYP_Encrypt(&data->hcryp, (uint32_t *)in_buf, in_len,
-				  (uint32_t *)out_buf, HAL_MAX_DELAY);
+	status = fn(&data->hcryp, in_buf, in_len, out_buf, HAL_MAX_DELAY);
 	if (status != HAL_OK) {
-		LOG_ERR("Encryption error");
+		LOG_ERR("Encryption/decryption error");
```
Split overall changes added to the driver (which will impact already supported IPs) and L4 support additions in distinct commits.

erwango commented on dts/arm/st/l4/stm32l462.dtsi:
```diff
@@ -11,7 +11,7 @@
 
 		aes: aes@50060000 {
-			compatible = "st,stm32-aes";
+			compatible = "st,stm32l4-aes";
```
Note: Alternatively, you can do:
```suggestion
			compatible = "st,stm32-aes", "st,stm32l4-aes";
```
This way, you benefit from `DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)` being defined and it can simplify things like:
```diff
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp)
#define DT_DRV_COMPAT st_stm32_cryp
#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
#define DT_DRV_COMPAT st_stm32_aes
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)    ===> No more required
+#define DT_DRV_COMPAT st_stm32l4_aes
```

lucasdietrich replied:
Consequently, I introduced USE_HAL_GENERIC and USE_HAL_L4 to differentiate the nature of the two HALs

lucasdietrich committed:
drivers: crypto: Refactor encryption and decryption functions for STM32 AES

This patch introduces a unified function pointer approach to handle
encryption and decryption operations for the STM32 AES accelerator.

- Replace separate `do_encrypt` and `do_decrypt` functions with a generic
`do_aes` function, using function pointers to AES HAL functions.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -47,11 +47,31 @@ LOG_MODULE_REGISTER(crypto_stm32);
-			      const uint8_t *src_buf, int src_len)
-{
-	int i;
-
```
Something is broken in your commit split.

lucasdietrich replied:
Fixed

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -45,7 +45,11 @@ LOG_MODULE_REGISTER(crypto_stm32);
 #endif
 
-struct crypto_stm32_session crypto_stm32_sessions[CRYPTO_MAX_SESSION];
+#define USE_HAL_GENERIC                                                                            \
+	(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) ||                                               \
+	 (DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes) && !DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)))
+
+#define USE_HAL_L4 (DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes))
```
Those 2 defines are not required. Checking `DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)` should be sufficient.

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -95,35 +114,21 @@ static int do_encrypt(struct cipher_ctx *ctx, uint8_t *in_buf, int in_len,
 
-static int do_decrypt(struct cipher_ctx *ctx, uint8_t *in_buf, int in_len,
-		      uint8_t *out_buf)
+#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
```
Nitpick: `!DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)` is not actually required in this commit

erwango commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -47,11 +47,31 @@ LOG_MODULE_REGISTER(crypto_stm32);
+ *
+ * @retval HAL_StatusTypeDef  HAL status of the operation.
+ */
+typedef HAL_StatusTypeDef (*hal_cryp_aes_op_func_t)(CRYP_HandleTypeDef *hcryp, uint8_t *in_data,
```
```suggestion
typedef status_t (*hal_cryp_aes_op_func_t)(CRYP_HandleTypeDef *hcryp, uint8_t *in_data,
```

erwango commented on drivers/crypto/crypto_stm32_priv.h:
```diff
@@ -8,6 +8,12 @@
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
+#define STM32_CRYPTO_CONFIG_TYPEDEF            CRYP_InitTypeDef
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_cryp) || DT_HAS_COMPAT_STATUS_OKAY(st_stm32_aes)
```
`#else` is sufficient here

erwango commented on drivers/crypto/crypto_stm32_priv.h:
```diff
@@ -8,6 +8,12 @@
 #define ZEPHYR_DRIVERS_CRYPTO_CRYPTO_STM32_PRIV_H_
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32l4_aes)
+#define STM32_CRYPTO_CONFIG_TYPEDEF            CRYP_InitTypeDef
```
#define crypt_config_t

erwango reviewed (changes requested):
Thanks for these changes, looks great.

Last couple of nitpicks, otherwise (and we'll be good to go):
- please update others L4 aes dts nodes
- at least one commit message needs rewording (the one with USE_HAL ...)
- and finally, we'll need to ensure this new code is compiled in CI, so there should be at least one matching in-tree board enabling this node.

lucasdietrich commented:
I believe I have addressed most of the points.

However, to build the crypto sample on a board with a SoC that includes the ST, STM32L4-AES driver, I initially chose the `nucleo_l496zg`. However, it appears that this soc does not have an AES accelerator at all. According to the reference manual (RM0351), in the section "AES hardware accelerator (AES)," only the STM32L486/L4A6 include the AES accelerator.

The compilation actually fails because the header file `stm32cube/stm32l4xx/soc/stm32l496xx.h` from the STM32 HAL does not contain the "AES" constant required to enable the AES HAL.

I decided to check other SoCs:
- STM32L422xx (DS12470 Rev 6): AES accelerator is present.
- STM32L4P5 (DS12903 Rev 3): There is no clear mention of the AES device (surprisingly, there is a line for the AES device in the power consumption table).


Unless I am mistaken, I believe we should restrict the `st,stm32l4-aes` driver to STM32L462 and STM32L422xx SoCs.

As a result, there is currently no board in the Zephyr tree that uses an SoC with the STM32L4-AES accelerator. Furthermore, I am not aware of any development kit (DK) or known board that includes an STM32L422 or STM32L462.

lucasdietrich commented:
> As a result, there is currently no board in the Zephyr tree that uses an SoC with the STM32L4-AES accelerator. Furthermore, I am not aware of any development kit (DK) or known board that includes an STM32L422 or STM32L462.

I just noticed that the STM32L4A6 SoC is present in the Zephyr tree. Should I add the AES node to the DTS for this SoC?

Morevoer, if a board for the [NUCLEO-L4A6ZG](https://www.st.com/en/evaluation-tools/nucleo-l4a6zg.html) could be added to the tree, it could serve as a candidate for "stm32l4-aes" testing in the CI.

lucasdietrich committed:
drivers: crypto: Add support for STM32L4 AES accelerator

This patch completes the addition of support for the STM32L4 AES
accelerator by introducing conditional handling for different STM32 AES
HAL variants. Key changes include:

- Created device tree bindings `st,stm32l4-aes` for STM32L4 AES
- Replaced `copy_reverse_words` with `copy_words_adjust_endianness`
to handle endianness conversion for different variants.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

lucasdietrich committed:
drivers: crypto: Fix pointer type warnings in STM32 AES driver

This patch resolves compiler warnings related to mismatched pointer types
between the STM32L4 and generic STM32 AES HAL by introducing CAST_VEC
macro.

Fix github CI warning

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

lucasdietrich committed:
dts: stm32l4: Update AES node for stm32l4 series

The stm32l4 devices were previously assigned the generic STM32 AES driver,
which turned out to be incompatible with the stm32l4 series. This commit
updates the nodes to use the new driver specifically designed for this
series.

Add missing node for stm32l4a6, stm32l4q5, stm32l4s5 and stm32l486 socs.

It appears stm32l4p5 and stm32l496 socs do not have the AES accelerator
present, so the nodes are removed from the dts files.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

lucasdietrich committed:
dts: Enable AES node in nucleo_l4a6zg.dts

Activate AES node in nucleo_l4a6zg.dts to enable testing of the
st,stm32l4-aes driver in CI.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

erwango commented on samples/drivers/crypto/sample.yaml:
```diff
@@ -51,6 +51,22 @@ tests:
         - ".*: CBC mode DECRYPT - Match"
         - ".*: CTR mode ENCRYPT - Match"
         - ".*: CTR mode DECRYPT - Match"
+  sample.drivers.crypto.stm32l4:
+    tags: crypto
+    filter: dt_compat_enabled("st,stm32l4-aes")
+    integration_platforms:
+      - nucleo_l4a6zg
+    harness: console
+    harness_config:
+      type: multi_line
+      regex:
+        - ".*: Cipher Sample"
+        - ".*: ECB mode ENCRYPT - Match"
+        - ".*: ECB mode DECRYPT - Match"
+        - ".*: CBC mode ENCRYPT - Match"
+        - ".*: CBC mode DECRYPT - Match"
+        - ".*: CTR mode ENCRYPT - Match"
+        - ".*: CTR mode DECRYPT - Match"
```
This is not required. What you just need to do is enable aes node directly in `nucleo_l4a6zg.dts` and then it will match filter of already existing test config `sample.drivers.crypto.stm32` thanks to `dt_compat_enabled("st,stm32-aes")`.

erwango reviewed (changes requested):
One very last fix and we'll be good to go!

erwango reviewed (approved):
Approved, Thanks for your time!

lucasdietrich commented:
You're welcome, just wanted to check if there's any update on merging the PR. Please let me know if I can help further.
